qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+            elif driven_by == "implicit":
+                W_inew = self.Hamiltonian_manager.implicit_matrix_product(V_new)
+            else:
+                raise NotImplementedError
+            W_ik = np.c_[W_ik, W_inew]
+
+            # TODO: Don't need to reduce full matrix at each step. Just new columns/rows
+            # Compute rank computes partial update to the projected Hamiltonian S_k
+            if (k == 1) or (
+                restart == 1
+            ):  # If first iterate (or following a restart), need to compute full S_k explicitly
+                S_ik = np.dot(V_ik.T, W_ik)
+                restart = 0
+            else:  # Else, append new rows & columns
+                S_inew_c = np.dot(V_ik[:, :-n_newvecs].T, W_inew)
+                S_ik = np.c_[S_ik, S_inew_c]
+                S_inew_r = np.c_[np.dot(V_inew.T, W_ik[:, :-n_newvecs]), np.dot(V_inew.T, W_inew)]
+                S_ik = np.r_[S_ik, S_inew_r]
+            # Reduce contributions and form new S_k
+            S_k = None  # Initialize
+            if self.rank == 0:
+                S_k = np.zeros((dim_S, dim_S), dtype="float")
+            self.comm.Reduce([S_ik, MPI.DOUBLE], [S_k, MPI.DOUBLE])
+            if self.rank == 0:  # Master diagonalizes S_k
+                L_k, Y_k = np.linalg.eigh(S_k)
+                L_k = np.array(L_k[:n_eig], dtype="float")
+                Y_k = np.array(Y_k[:, :n_eig], dtype="float")
+            else:  # Pre-allocate space to receive
+                L_k = np.zeros(n_eig, dtype="float")
+                Y_k = np.zeros((dim_S, n_eig), dtype="float")
+            # Broadcast smallest n_eig estimates to all ranks to compute residuals
+            self.comm.Bcast([L_k, MPI.DOUBLE], root=0)
+            self.comm.Bcast([Y_k, MPI.DOUBLE], root=0)
+
+            n_newvecs = 0  # Initialize counter; no. of new vectors added to trial subspace
+            X_ik = np.dot(V_ik, Y_k)  # Pre-compute Ritz vectors (V_ik updated each iteration)
+            converged = True  # Reset
+            for j in range(n_eig):  # TODO: Remove converged eigenpairs
+                # Grab jth eigenpair, each rank computes local portion of jth residual vector
+                l_j, y_j = L_k[j], Y_k[:, j]
+                r_ij = np.dot(W_ik, y_j) - l_j * X_ik[:, j]
+                r_j = np.zeros(n, dtype="float")
+                # Gather residual on each rank and compute its norm
+                self.comm.Allgatherv(
+                    [r_ij, MPI.DOUBLE], [r_j, self.distribution, self.offsets, MPI.DOUBLE]
+                )
+                res = np.linalg.norm(r_j)
+                if self.rank == 0:
+                    print(f"||r_k||: {res}")
+                if res > eps:  # If ||r_j|| > eps, jth eigenpair hasn't converged
+                    converged = False
+                    if self.rank == 0:
+                        print(f"Eigenvalue {j}: not converged, preconditioning next trial vector")
```

Comment:
To simplify the code, create a function call `print_master` who do this `if rank== MASTER_RANK
print)